src/compiler/crystal/syntax/parser.cr

```diff
@@ -3534,6 +3526,22 @@ module Crystal
         end
       end
 
+      if name.ends_with?('=') && !name.in?("<=", "==", ">=", "!=", "===")
+        if name != "[]=" && (args.size != 1 || found_splat || found_double_splat)
+          raise "setter method '#{name}' must have exactly one positional argument"
+        elsif name == "[]=" && args.size < 1
+          raise "setter method '#{name}' must have at least one positional argument"
```

Comment:
It looks like `def []=(x, *x)` would be valid according to these conditions.

Reply:
`def []=(x, *y)` is legal because the array indices can be empty (i.e. `StaticArray#[]=`, until recently). `def []=(*x : Object)` is also legal. `def []=(*x)` isn't and I'll fix that soon.